### ISSUE TITLE
ENH: Type stubs

### ISSUE THREAD

**Issue by ctriley** [geopandas/geopandas#3091]
#### Is your feature request related to a problem?

I want to have type stubs when using geopandas

https://typing.readthedocs.io/en/latest/source/stubs.html

#### Describe the solution you'd like

I'd like geopandas to include type stubs.

#### API breaking implications

[this should provide a description of how this feature will affect the API]

It will allow type annotations when using geopandas.

**Comment by martinfleis** [comment]
We are slowly adding type hints to the code base but it hasn't been a priority (ref #1974).

I don't think there'll be an appetite in the team to work on stub files. If something, we will be adding type hints directly but I can't promise anything on when geopandas will be fully typed.

**Comment by m-richards** [comment]
From your perspective as a user, is there any benefit to having stub files compared with having inline, [PEP561](https://peps.python.org/pep-0561) compliant types coverage with a `py.typed` file distributed? As it stands the latter is the current direction as Martin says.

**Comment by estyrke** [comment]
Chiming in here - from my perspective it doesn't matter if its via stubs or inline (but inline seems preferable when browsing the code). My pain right now is having to write `cast(gpd.GeoDataFrame, my_gdf.some_pandas_operation_which_I_know_will_return_a_gdf())` all the time to prevent the type checker from complaining. If that can be solved I don't really care how it's done. :)

I am using vs code which has bundled stubs for Pandas, which means that `DataFrame.__getitem__` is correctly typed depending on the `key` for example. Having correct type hints for `GeoDataFrame.__getitem__` would also come pretty high on my wish list.

**Comment by m-richards** [comment]
We would certainly like to improve the typing situation, and methods which definitely return a geodataframe are low hanging fruit.

> I am using vs code which has bundled stubs for Pandas, which means that `DataFrame.__getitem__` is correctly typed depending on the `key` for example. Having correct type hints for `GeoDataFrame.__getitem__` would also come pretty high on my wish list.

The unfortunate thing with `__getitem__` is that it is dynamically typed from a geopandas perspective. So even if we had type annotations, if `key` is a string the return type is `Series|GeoSeries` and if `key` is a list of column names then the return type is `DataFrame|GeoDataFrame` which is likely not going to save you many `cast`s

**Comment by estyrke** [comment]
I think it would save me at least some casts; at least when I wrote that comment the issue was the ambiguity between `*Series` and `*DataFrame`. But I guess it will never be perfect unless it's possible to statically determine which column(s?) contains geometry.

**Comment by hamdanal** [comment]
Hello all, I added complete geopandas stubs to typeshed. Please checkout https://pypi.org/project/types-geopandas/ and feel free to report any bugs/suggestions to the typeshed project.

Also, if you are not aware already, I added complete shapely stubs to typeshed a while back (see https://pypi.org/project/types-shapely/). I also added a `py.typed` marker file to folium and branca which means they are considered typed by mypy.

This issue can be closed.

**Comment by m-richards** [comment]
Hi @hamdanal, thanks for this, I will definitely take a look. I would suggest though that it's a good idea to discuss this kind of contribution with the project you're contributing stubs for in advanced of them being released. There has been an effort to develop type annotations directly within geopandas over quite a period of time now. 

I appreciate the desire to move faster than we've been merge type stubs PRs here directly, however I think we would generally have appreciated the window to review annotations (even on typeshed) before they're merged, let alone before they're published on pypi.

**Comment by hamdanal** [comment]
> I would suggest though that it's a good idea to discuss this kind of contribution with the project you're contributing stubs for in advanced of them being released.

Usually the discussion happens to see if the maintainers of the project would prefer to own the stubs or to include inline types instead. Since both these topics have been discussed, I didn't feel that I would've contributed anything new to the discussion. It was decided to have inline typing but there is not much progress for almost 4 years now (which is totally fine, this is how open source works). I see the typeshed stubs as a temporary solution for users until inline typing is finished.

> There has been an effort to develop type annotations directly within geopandas over quite a period of time now.

This is great. The moment you are happy with the level of typing in your project and you include the `py.typed` marker, typeshed automatically deprecates and removes their stubs. Having stubs in typeshed does not prohibit adoption of inline typing. On the contrary, it provides a solution to users who need typing now and relieves you from the pressure of adding stubs right away to do it on your own pace and however your time permits. It also helps in that you, the maintainers, can use the types in typeshed when annotating your code base.

> I appreciate the desire to move faster than we've been merge type stubs PRs here directly, however I think we would generally have appreciated the window to review annotations (even on typeshed) before they're merged, let alone before they're published on pypi.

It is not typeshed policy to require any kind of commitment, in reviews or otherwise, from upstream projects. You are free to propose changes to the stubs as is everyone. typeshed maintainers usually appreciate input from experts on upstream project, but they do not require it. I am sure any suggestions you have will be welcome there.

*Note that I am not a typeshed maintainer, I am just a regular contributor there.*

Finally, I want to say that your and your fellow maintainers work on geopandas is highly appreciated. Thank you all for the time you spend on maintaining and improving it.

**Comment by m-richards** [comment]
Thanks @hamdanal that's all quite useful to know to get a bit more context about how the typeshed process works. 

Question from a position of ignorance, is there anything from the typeshed stubs that you feel wouldn't make sense in inline stubs / would be done differently in inline stubs? We have a number of long standing PRs with contributions for type hints, but it may be good to cross reference them with the work you've done.

From the top of my head something that might come to mind as different is situations where we have a transitioning type signature e.g. `GeoDataFrame.set_geometry` has an argument `drop: bool|None` which is typed as such inline, because the default of `None` shows a warning and aliases to `False`. But I could imagine in "user facing" type stubs that might be typed as just `bool`.

**Comment by hamdanal** [comment]
> Question from a position of ignorance, is there anything from the typeshed stubs that you feel wouldn't make sense in inline stubs / would be done differently in inline stubs?

Yes I do actually. I think long overloads can be cumbersome to have inline. They clutter the code and make it less readable. This is not the case in stubs as there is no runtime code that you need to read there. This is of course subjective, so it really depends on how you like to maintain your project.

> From the top of my head something that might come to mind as different is situations where we have a transitioning type signature e.g. `GeoDataFrame.set_geometry` has an argument `drop: bool|None` which is typed as such inline, because the default of `None` shows a warning and aliases to `False`. But I could imagine in "user facing" type stubs that might be typed as just `bool`.

I usually try to stick to runtime types, the documentation and warnings should be enough to signal deprecations. With that being said, you can use `@deprecated` (from `typing_extensions`) to mark certain overloads of a function as deprecated. Type checkers like mypy would then show a warning or an error to the user. Some IDEs like vscode with Pylance will also show a warning in the editor. In your case here, you'd create a deprecated overload with `drop: None` and a normal overload with `drop: bool`.
